internal/it/test_cluster.go

```diff
@@ -0,0 +1,301 @@
+	}
+}
+
+func (c *SingletonTestCluster) Launch(t testLogger) TestCluster {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cls != nil {
+		return c.cls
+	}
+	t.Logf("Launching the auto-shutdown test cluster: %s", c.name)
+	c.cls = c.launcher()
+	return c.cls
+}
+
+func StartNewClusterWithOptions(clusterName string, port, memberCount int) *DedicatedTestCluster {
+	ensureRemoteController(false)
+	config := xmlConfig(clusterName, port)
+	return rc.startNewCluster(memberCount, config, port)
+}
+
+func StartNewClusterWithConfig(memberCount int, config string, port int) *DedicatedTestCluster {
+	ensureRemoteController(false)
+	return rc.startNewCluster(memberCount, config, port)
+}
+
+func xmlConfig(clusterName string, port int) string {
+	return fmt.Sprintf(`
+        <hazelcast xmlns="http://www.hazelcast.com/schema/config"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://www.hazelcast.com/schema/config
+            http://www.hazelcast.com/schema/config/hazelcast-config-4.0.xsd">
+            <cluster-name>%s</cluster-name>
+            <network>
+               <port>%d</port>
+            </network>
+			<jet enabled="true" />
+        </hazelcast>
+	`, clusterName, port)
+}
+
+type viridianClusterInfo struct {
+	ID    string `json:"id"`
+	State string `json:"state"`
+}
+
+type keyValue map[string]any
+
+type Wrapper[T any] struct {
+	Content T
+}
+
+type ViridianAPI struct {
+	token string
+}
+
+func loginNewViridianAPI(ctx context.Context, apiKey, apiSecret string) *ViridianAPI {
+	a := &ViridianAPI{}
+	a.login(ctx, apiKey, apiSecret)
+	return a
+}
+
+func (a *ViridianAPI) login(ctx context.Context, apiKey, apiSecret string) {
+	req := map[string]any{
+		"apiKey":    apiKey,
+		"apiSecret": apiSecret,
+	}
+	res, err := doPost[keyValue, keyValue](ctx, "/customers/api/login", "", req)
+	if err != nil {
+		panic(err)
+	}
+	a.token = res["token"].(string)
+}
+
+func (a *ViridianAPI) CreateCluster(ctx context.Context, name string) (viridianClusterInfo, error) {
+	req := map[string]any{
+		"kubernetesClusterId": 1,
+		"clusterTypeId":       6,
+		"name":                name,
+		"planName":            "SERVERLESS",
+	}
+	res, err := doPost[keyValue, viridianClusterInfo](ctx, "/cluster", a.token, req)
+	if err != nil {
+		return viridianClusterInfo{}, err
+	}
+	return res, nil
+}
+
+func (a *ViridianAPI) DeleteCluster(ctx context.Context, clusterID string) error {
+	return doDelete(ctx, "/cluster/"+clusterID, a.token)
+}
+
+func (a *ViridianAPI) ListClusters(ctx context.Context) ([]viridianClusterInfo, error) {
+	res, err := doGet[Wrapper[[]viridianClusterInfo]](ctx, "/cluster", a.token)
+	if err != nil {
+		return nil, err
+	}
+	return res.Content, nil
+}
+
+func makeUrl(path string) string {
+	path = strings.TrimLeft(path, "/")
+	path = "/" + path
+	return os.Getenv(envAPIBaseURL) + path
+}
+
+func doPost[Req, Res any](ctx context.Context, path, token string, request Req) (res Res, err error) {
```

Comment:
I have a question. I guess we did not want to expose the `doPost`, `doGet` etc. functions in `internal/viridian/api.go`.  I am not sure if we intend to change the code in these functions but does it make sense to create a client package that will expose these functions so that the client package can be used in the prod code and tests?

Reply:
We could do that, but this is some trivial code that is supposed to be very simple and should be used just to run the tests.. I think it would be OK to duplicate them in order to not create a dependency. This way we will be able make changes to the API in internal/viridian without updating the tests.